docs/pages/pmd/projectdocs/decisions/adr-3.md

```diff
@@ -0,0 +1,198 @@
+
+The actual API development and marking some part of the API as internal or add new API is an ongoing task,
+that will need to be done everytime. We won't just define an API and then are done with it.
+The API will change as new features want to be implemented.
+
+This decision document aims to document principles and guidelines that are used for PMD development.
+
+## Decision
+
+### Semantic Versioning
+
+PMD and all its modules are versioned together. PMD uses [Semantic Versioning 2.0.0](https://semver.org/spec/v2.0.0.html).
+This means, that each PMD version consists of MAJOR.MINOR.PATCH components:
+
+* MAJOR version is incremented for incompatible API changes
+* MINOR version is incremented for added functionality in a backwards compatible way
+* PATCH version is incremented for backward compatible bug fixes
+
+Additional labels for release candidates might be used.
+
+Incompatible API changes shouldn't be introduced lightly. See
+[FAQ: If even the tiniest backward incompatible changes to the public API require a major version bump, won’t I end up at version 42.0.0 very rapidly?](https://semver.org/spec/v2.0.0.html#if-even-the-tiniest-backward-incompatible-changes-to-the-public-api-require-a-major-version-bump-wont-i-end-up-at-version-4200-very-rapidly).
+
+### Project structure and Java base packages names
+
+PMD is mainly developed in the Java programming language. The build tool is Maven and the PMD build consists
+of several maven modules.
+
+* Classes belonging to one module should be in the same package (or sub-packages).
+* Given a package name, it should be easy to figure out to which module this package belongs. There is a 1:1 mapping
+  between maven module and package. This rule helps to find the source code for any fully qualified (Java) class name.
+* Two modules must not define the same packages. That means, it is not allowed that any given package spans more than
+  one module. Otherwise, the mapping between module and package wouldn't be unambiguous.
+* The base package for all PMD source code is `net.sourceforge.pmd`. There are many different sub packages.
+* The core module `pmd-core` uses directly the base package as the only module. All other modules must use
+  specific sub packages.
+* Language modules use the base package `net.sourceforge.pmd.lang.<language id>`.
+  E.g. `pmd-java` uses the package `net.sourceforge.pmd.lang.java`.
+* All other modules use the base package `net.sourceforge.pmd.<module>`,
+  E.g. `pmd-cli` uses the package `net.sourceforge.pmd.cli`.
+
+### Criteria for public API
+
+Public API is
+
+* API needed to execute PMD analysis
+  * Renderers
+  * RuleSet XML Schema
+  * Configuration
+  * Ant Tasks
+* API needed to implement custom rules
+  * AST structure and classes of languages (incl. AST structure for XPath rules)
+  * XPath functions
+  * Language Symbol Table / Metrics / Type Resolution (Not the implementation)
+
+**Not** public API is
+
+* Anything in packages `internal` and `impl`
```

Comment:
I understood the packages named `impl` as public API of a module that would only be relevant to a person implementing a new language module. As in, pmd-java can depend on `impl` packages of PMD core, but not on internal packages ideally.

But I'm fine with keeping it that way, as in treat `impl` as internal API. The distinction would only be relevant if we want to support the use case of people implementing their own language modules outside of our monorepo. While I think this use case should be supported eventually, we can also for now bar it off and decide later.

Reply:
Yes, I think we should stay pragmatic for now and keep - at least for 7.0.0 - `impl` as private. If you compare this to linux kernel (module) development, then the userpace API is public API and needs to stay compatible, but anything else - including the API to implement modules - is private and can change with any patch version. That's why a module developed outside the kernel tree will only be compatible against the kernel version, that it is compiled for and whenever it is compiled against a newer version, one might need to adapt it to changes done in the kernel. That's why usually the goals is to have the modules integrated in the kernel tree. This is similar to developing a new language for PMD (which is a new module) - if it is developed outside our monorepo, then there are no guarantees, whether it will work with a newer PMD version now. I believe, this is doable - given that I don't know of anyone who is developing a language module outside the monorepo (ok, that doesn mean anything...).
But pragmatically, I would simply make `impl` for now private, until we get some bug reports about something has been broken. And only then, we should think about make it a public API.

> I understood the packages named impl as public API of a module that would only be relevant to a person implementing a new language module. As in, pmd-java can depend on impl packages of PMD core, but not on internal packages ideally.

That's a good summary, I'll try to incorporate this into the text.